attachment widget: data defined button for document type

3nids opened the pull request:
not fully tested yet

3nids committed:
attachment widget: data defined button for document type

3nids commented:
@nyalldawson do we have any other option to add properties that are related to QGIS enums? You would rather go for text values and match to the enum or use integers as I did?

nyalldawson commented:
> You would rather go for text values and match to the enum or use integers as I did?

@3nids I think string parsing is nicer. It's easier (at least for english speakers!) to understand the expressions if they use string return values, and it makes things more future proof if we ever need to modify or rearrange that enum in future. There's also the risk that an expression can return ANY int, which won't cast nicely across to an existing enum value, leaving things in a undefined state.

I'd model this on something like QgsSymbolLayer::PropertyPlacement - with a function that parses strings to  QgsExternalResourceWidget::DocumentViewerContent.

nyalldawson commented on src/gui/editorwidgets/qgsexternalresourcewidgetwrapper.cpp:
```diff
@@ -79,15 +79,29 @@ bool QgsExternalResourceWidgetWrapper::valid() const
-      mQgsWidget->setDefaultRoot( path );
+    if ( mPropertyCollection.isActive( QgsEditorWidgetWrapper::RootPath ) )
+    {
+      QgsExpressionContext expressionContext( QgsExpressionContextUtils::globalProjectLayerScopes( layer() ) );
```
I'd suggest moving the construction of the expression context up from this scope so that the same context can be reused by both properties

3nids replied:
fixed

nyalldawson commented on src/gui/editorwidgets/qgsexternalresourceconfigdlg.cpp:
```diff
@@ -107,6 +111,16 @@ void QgsExternalResourceConfigDlg::rootPathPropertyChanged()
+
+  mDocumentViewerContentExpression->setVisible( prop.isActive() );
+  //mDocumentViewerContent->setVisible( !prop.isActive() );
+  mDocumentViewerContentComboBox->setEnabled( !prop.isActive() );
```
There was (unused) code in the old QgsDataDefinedButton class which automatically handled this use case. I wonder if it's worth resurrecting and copying to QgsPropertyOverrideButton? (see https://github.com/qgis/QGIS/blob/release-2_18/src/gui/qgsdatadefinedbutton.h#L180)

3nids replied:
fixed. I added for visible and enable states (with option to have it the other way around)

nyalldawson commented on src/ui/editorwidgets/qgsexternalresourceconfigdlg.ui:
```diff
@@ -435,6 +435,20 @@
+          </item>
+          <item row="0" column="3">
+           <widget class="QLineEdit" name="mDocumentViewerContentExpression">
+            <property name="enabled">
```
I think this should be read only, not disabled?

3nids replied:
fixed

3nids commented:
@nyalldawson thanks for the review. 

I completed QgsPropertyOverrideButton with more sibling possibilities, is it fine for you?

Also, I think I have an issue with the state of the property button: sibling widgets are not displayed properly on first show (although I call the corresponding slots in the 3 registerXXXwidget methods) and it seems the activated signal is only emitted when you click activate/deactivate menu entry but not when you first choose a field for instance. Do you confirm?

nyalldawson commented on src/gui/editorwidgets/qgsexternalresourcewidgetwrapper.cpp:
```diff
@@ -79,15 +79,41 @@ bool QgsExternalResourceWidgetWrapper::valid() const
 void QgsExternalResourceWidgetWrapper::setFeature( const QgsFeature &feature )
 {
-  if ( mQgsWidget && mPropertyCollection.isActive( QgsEditorWidgetWrapper::RootPath ) )
+  if ( mQgsWidget && ( mPropertyCollection.isActive( QgsEditorWidgetWrapper::RootPath )
```
You can use mPropertyCollection.hasActiveProperties() instead here

nyalldawson commented on src/gui/qgspropertyoverridebutton.cpp:
```diff
@@ -78,6 +80,13 @@ QgsPropertyOverrideButton::QgsPropertyOverrideButton( QWidget *parent,
 
+QgsPropertyOverrideButton::~QgsPropertyOverrideButton()
+{
+  mEnabledWidgets.clear();
```
These clears aren't needed - they don't do anything

3nids commented on src/gui/editorwidgets/qgsexternalresourceconfigdlg.cpp:
```diff
@@ -253,3 +265,12 @@ void QgsExternalResourceConfigDlg::setRootPathExpression( const QgsProperty &pro
   mRootPathExpression->setText( property.valueAsString( ctx ) );
```
@nyalldawson @m-kuhn this was not working in current code (to get the expression in the line edit). It seems `QgsProperty::valueAsString` can't return a value. Any idea?

3nids replied:
@nyalldawson @m-kuhn this was not working in current code (to get the expression in the line edit). It seems `QgsProperty::valueAsString` can't return a value. Any idea?

3nids commented on src/gui/qgspropertyoverridebutton.cpp:
```diff
@@ -119,6 +122,10 @@ void QgsPropertyOverrideButton::init( int propertyKey, const QgsProperty &proper
+
+  checkCheckedWidgets( isActive() );
+  enableEnabledWidgets( isActive() );
+  visibleVisibleWidgets( isActive() );
```
@nyalldawson this had not effect on the initial state. I don't understand why it's not correctly setup.

3nids commented:
@nyalldawson I could fix part of the issues. I added another register method for widgets (lineedits) to display the expression, it works as expected. 
The only issue left is when using the menu entry to give a field: the sibling widgets are not updarted. See the video: https://youtu.be/HJ2tWBy8_no
It seems the property is active (as shown in the debugger) while it is not at that moment in reality, hence `setActivePrivate` does not emit the signal (because it thinks the property is already active).

Also I removed the blockSignal as I didn't understand their utility. I left them commented.

nyalldawson commented on src/gui/qgspropertyoverridebutton.cpp:
```diff
@@ -725,22 +733,46 @@ void QgsPropertyOverrideButton::enableEnabledWidgets( bool enable )
+  }
+}
+
+void QgsPropertyOverrideButton::setActive( bool active, bool forceSignal )
```
I'm not a huge fan of these "forceSignal" type arguments - they tend to quickly turn into messy corner cases. 

Here's an alternative idea - why not require that that init is called before registering widgets? Then the various widget registration functions can perform the disabling/checking/etc immediately upon registration.

nyalldawson commented:
> It seems the property is active (as shown in the debugger) while it is not at that moment in reality, hence setActivePrivate does not emit the signal (because it thinks the property is already active).

You've also got invalid properties to handle here. That's what all the various !mProperty checks are doing. I suspect that's where this is getting caught
